File: javascript/23.merge-k-sorted-lists.js
/*
 * @lc app=leetcode id=23 lang=javascript
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
  if (lists.length === 0 ||
      (!lists.some(v => v !== null))) return null

  let head = currNode = new ListNode()
  // lists = lists.reduce(v => v !== null)
  // iterate until no nodes remain
  while (lists.some(v => v !== null)) {
    // compare each node val in each list and add min
    let selectedIndex = -1
    currNode.val = Number.MAX_SAFE_INTEGER
    for (let [index, node] of lists.entries()) {
      if (!node) continue
      if (node.val < currNode.val) {
        currNode.val = node.val
        selectedIndex = index
      }
    }
    // increment pointer in chosen list
    lists[selectedIndex] = lists[selectedIndex] ? lists[selectedIndex].next : null
    if (!lists.some( v => v !== null)) break
    currNode.next = new ListNode()
    currNode = currNode.next
  }
  return head
};
// @lc code=end

File: cpp/23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        // iterate over every linked list
        // select the smallest int
        // move position in that linked list
        // O(n^2) time O(n) space
        ListNode *currNode = new ListNode{};
        ListNode *mergeHead = new ListNode{0, currNode};
        while (true)
        {
            int lowestNum = pow(10, 5);
            int lowestNumIndex = -1;
            for (int i = 0; i < lists.size(); i++)
            {
                if (lists[i] == nullptr)
                {
                    lists.erase(lists.begin() + i);
                    i--;
                    continue;
                }
                if (lists[i]->val <= lowestNum)
                {
                    lowestNum = lists[i]->val;
                    lowestNumIndex = i;
                }
            }
            // check if there were any non-null linked lists left
            if (lowestNumIndex > -1)
            {
                currNode->next = new ListNode(lowestNum);
                currNode = currNode->next;
                lists[lowestNumIndex] = lists[lowestNumIndex]->next;
            }
            else
            {
                break;
            }
        }
        return mergeHead->next->next;
    }
};
// @lc code=end
